File: packages/mobile/android/src/main/kotlin/app/abto/sdk/AbtoTransport.kt
package app.abto.sdk

import java.net.HttpURLConnection
import java.net.URI
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.TimeUnit

internal fun readBoundedResponse(
    input: InputStream,
    maxBytes: Int = 64 * 1024,
): String {
    val output = ByteArrayOutputStream(minOf(maxBytes, 8 * 1024))
    val chunk = ByteArray(8 * 1024)
    var total = 0
    while (true) {
        val read = input.read(chunk)
        if (read < 0) break
        total += read
        if (total > maxBytes) throw IOException("[abto] collector response exceeded $maxBytes bytes.")
        output.write(chunk, 0, read)
    }
    return output.toString(Charsets.UTF_8)
}

/**
 * Batch transport that queues events and flushes on `batchSize` or a timer.
 * Network operations always run on a dedicated thread to avoid Android main-thread restrictions
 * (`NetworkOnMainThreadException`), and failures are never thrown to the host app, matching the Browser SDK contract.
 */
class AbtoTransport(private val config: AbtoConfig) {
    private data class QueuedEvent(
        val event: Map<String, Any?>,
        val firstQueuedAtMs: Long = System.currentTimeMillis(),
        var attempts: Int = 0,
    )

    private val lock = Any()
    private val buffer = ArrayDeque<QueuedEvent>()
    private val executor = Executors.newSingleThreadScheduledExecutor { runnable ->
        Thread(runnable, "abto-transport").apply { isDaemon = true }
    }
    private var scheduledFlush: ScheduledFuture<*>? = null
    private var flushRunning = false
    private var flushRequested = false
    private val pendingCompletions = ArrayDeque<Runnable>()

    fun enqueue(event: Map<String, Any?>) {
        synchronized(lock) {
            buffer.addLast(QueuedEvent(event))
            while (buffer.size > MAX_BUFFER) buffer.removeFirst()
            if (buffer.size >= config.batchSize) {
                requestFlushLocked(0)
            } else {
                scheduleFlushLocked(config.flushIntervalMs)
            }
        }
    }

    fun flush(onComplete: Runnable? = null) {
        synchronized(lock) {
            onComplete?.let(pendingCompletions::addLast)
            requestFlushLocked(0)
        }
    }

    private fun runScheduledFlush() {
        val completions: List<Runnable>
        synchronized(lock) {
            scheduledFlush = null
            if (flushRunning) {
                flushRequested = true
                return
            }
            flushRunning = true
            flushRequested = false
            completions = pendingCompletions.toList()
            pendingCompletions.clear()
        }

        val retryDelayMs = drainAndSend()

        synchronized(lock) {
            flushRunning = false
            val requestImmediately = flushRequested || pendingCompletions.isNotEmpty()
            flushRequested = false
            when {
                requestImmediately -> requestFlushLocked(0)
                retryDelayMs != null -> scheduleFlushLocked(retryDelayMs)
                buffer.isNotEmpty() -> requestFlushLocked(0)
            }
        }
        completions.forEach { completion ->
            try {
                completion.run()
            } catch (_: Exception) {
                // Telemetry completion callbacks must not terminate the transport worker.
            }
        }
    }

    private fun drainAndSend(): Long? {
        val batch: List<QueuedEvent>
        synchronized(lock) {
            if (buffer.isEmpty()) return null
            val count = minOf(buffer.size, config.batchSize, MAX_BATCH_SIZE)
            batch = List(count) { buffer.removeFirst() }
            batch.forEach { it.attempts += 1 }
        }

        val events = batch.map(QueuedEvent::event)
        val body = AbtoJson.encode(mapOf("batch" to events)).toByteArray(Charsets.UTF_8)
        val retryBatch = try {
            val connection = URI(config.endpoint).toURL().openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                connection.doOutput = true
                connection.connectTimeout = 5_000
                connection.readTimeout = 5_000
                connection.setRequestProperty("content-type", "application/json")
                connection.setRequestProperty("authorization", "Bearer ${config.projectKey}")
                connection.outputStream.use { it.write(body) }
                val status = connection.responseCode
                if (isTransientStatus(status)) {
                    batch
                } else if (status >= 400) {
                    emptyList()
                } else {
                    val responseBody = connection.inputStream.use { readBoundedResponse(it) }
                    val eventIds = events.mapNotNull { it["event_id"] as? String }
                    val retryIds = retryEventIds(responseBody, eventIds)
                    if (retryIds == null) {
                        batch
                    } else {
                        batch.filter {
                            it.event["event_id"] !is String || it.event["event_id"] in retryIds
                        }
                    }
                }
            } finally {
                connection.disconnect()
            }
        } catch (_: Exception) {
            batch
        }

        val nowMs = System.currentTimeMillis()
        val eligibleRetryBatch = retryBatch.filter {
            retryEligible(it.attempts, it.firstQueuedAtMs, nowMs)
        }
        if (eligibleRetryBatch.isNotEmpty()) {
            synchronized(lock) {
                // Cap the buffer so a dead endpoint cannot grow memory without bound.
                eligibleRetryBatch.asReversed().forEach { buffer.addFirst(it) }
                while (buffer.size > MAX_BUFFER) buffer.removeLast()
            }
            return retryDelayMs(eligibleRetryBatch)
        }
        return null
    }

    private fun requestFlushLocked(delayMs: Long) {
        if (flushRunning) {
            flushRequested = true
            return
        }
        if (delayMs == 0L && scheduledFlush != null) {
            scheduledFlush?.cancel(false)
            scheduledFlush = null
        }
        scheduleFlushLocked(delayMs)
    }

    private fun scheduleFlushLocked(delayMs: Long) {
        if (scheduledFlush != null) return
        scheduledFlush = executor.schedule({ runScheduledFlush() }, delayMs, TimeUnit.MILLISECONDS)
    }

    private fun retryDelayMs(batch: List<QueuedEvent>): Long {
        val attempt = batch.maxOfOrNull(QueuedEvent::attempts) ?: 1
        val baseMs = maxOf(1, config.flushIntervalMs)
        val exponentialMs = minOf(MAX_RETRY_DELAY_MS, baseMs * (1L shl (attempt - 1)))
        val jitterBound = maxOf(1, exponentialMs / 2)
        return exponentialMs + ThreadLocalRandom.current().nextLong(jitterBound)
    }

    private companion object {
        const val MAX_BUFFER = 1000
        const val MAX_BATCH_SIZE = 100
        const val MAX_ATTEMPTS = 3
        const val MAX_EVENT_AGE_MS = 5 * 60 * 1000L
        const val MAX_RETRY_DELAY_MS = 60 * 1000L
        val ACK_RESULTS = setOf("ok", "warning", "drop")
        val RESULT_PATTERN = Regex(""""result"\s*:\s*"([^"]+)"""")

        fun isTransientStatus(status: Int): Boolean = status == 408 || status == 429 || status >= 500

        fun retryEligible(attempts: Int, firstQueuedAtMs: Long, nowMs: Long): Boolean =
            attempts < MAX_ATTEMPTS && nowMs - firstQueuedAtMs < MAX_EVENT_AGE_MS

        fun retryEventIds(responseBody: String, eventIds: List<String>): Set<String>? {
            if (!Regex(""""results"\s*:\s*\{""").containsMatchIn(responseBody)) return null
            return eventIds.filterTo(linkedSetOf()) { eventId ->
                val eventPattern = Regex(""""${Regex.escape(eventId)}"\s*:\s*\{""")
                val match = eventPattern.find(responseBody) ?: return@filterTo true
                val objectStart = responseBody.indexOf('{', match.range.first)
                val objectEnd = findObjectEnd(responseBody, objectStart)
                if (objectEnd == null) {
                    true
                } else {
                    val result = RESULT_PATTERN.find(responseBody.substring(objectStart, objectEnd + 1))
                        ?.groupValues
                        ?.get(1)
                    result !in ACK_RESULTS
                }
            }
        }

        fun findObjectEnd(json: String, start: Int): Int? {
            if (start !in json.indices || json[start] != '{') return null
            var depth = 0
            var inString = false
            var escaped = false
            for (index in start until json.length) {
                val ch = json[index]
                if (inString) {
                    when {
                        escaped -> escaped = false
                        ch == '\\' -> escaped = true
                        ch == '"' -> inString = false
                    }
                    continue
                }
                when (ch) {
                    '"' -> inString = true
                    '{' -> depth += 1
                    '}' -> {
                        depth -= 1
                        if (depth == 0) return index
                    }
                }
            }
            return null
        }
    }
}

File: packages/mobile/swift/Sources/AbtoApp/AbtoTransport.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

package func abtoRetryEventIDs(responseData: Data, eventIDs: [String]) -> Set<String>? {
    guard let decoded = try? JSONSerialization.jsonObject(with: responseData),
          let root = decoded as? [String: Any],
          let results = root["results"] as? [String: Any] else {
        return nil
    }
    let acknowledged = Set(["ok", "warning", "drop"])
    return Set(eventIDs.filter { eventID in
        guard let eventResult = results[eventID] as? [String: Any],
              let result = eventResult["result"] as? String else {
            return true
        }
        return !acknowledged.contains(result)
    })
}

package func abtoRetryEligible(
    attempts: Int,
    firstQueuedAt: Date,
    now: Date = Date()
) -> Bool {
    attempts < abtoMaxAttempts && now.timeIntervalSince(firstQueuedAt) < abtoMaxEventAge
}

/// Batch transport that queues events and flushes on `batchSize` or a timer.
/// Never throws, so telemetry failures cannot block the host app, matching the Browser SDK contract.
// Mutable state is confined to `queue`; immutable event payloads cross the
// queue boundary as Data rather than `[String: Any]`.
final class AbtoTransport: @unchecked Sendable {
    private let config: AbtoConfig
    private let queue = DispatchQueue(label: "abto.sdk.transport")
    private struct QueuedEvent: Sendable {
        let data: Data
        let firstQueuedAt: Date
        var attempts: Int
    }

    private var buffer: [QueuedEvent] = []
    private var timer: DispatchSourceTimer?
    private let session: URLSession

    init(config: AbtoConfig, session: URLSession = .shared) {
        self.config = config
        self.session = session
    }

    func enqueue(_ event: [String: Any]) {
        guard let encodedEvent = try? JSONSerialization.data(withJSONObject: event) else {
            return
        }
        queue.async {
            self.buffer.append(QueuedEvent(data: encodedEvent, firstQueuedAt: Date(), attempts: 0))
            // Cap at production time so a dead endpoint cannot grow memory without bound.
            // On overflow the oldest events go first: recent ones are more useful.
            if self.buffer.count > abtoMaxBufferedEvents {
                self.buffer.removeFirst(self.buffer.count - abtoMaxBufferedEvents)
            }
            if self.buffer.count >= self.config.batchSize {
                self.flushLocked()
            } else {
                self.armTimerLocked()
            }
        }
    }

    func flush(completion: (@Sendable () -> Void)? = nil) {
        queue.async {
            self.flushLocked(completion: completion)
        }
    }

    private func armTimerLocked(delay: TimeInterval? = nil) {
        guard timer == nil else { return }
        let t = DispatchSource.makeTimerSource(queue: queue)
        t.schedule(deadline: .now() + (delay ?? config.flushInterval))
        t.setEventHandler { [weak self] in
            self?.timer = nil
            self?.flushLocked()
        }
        t.resume()
        timer = t
    }

    private func flushLocked(completion: (@Sendable () -> Void)? = nil) {
        timer?.cancel()
        timer = nil
        guard !buffer.isEmpty else {
            completion?()
            return
        }
        let count = min(buffer.count, config.batchSize, abtoMaxBatchSize)
        var attemptedBatch = Array(buffer.prefix(count))
        for index in attemptedBatch.indices {
            attemptedBatch[index].attempts += 1
        }
        let batch = attemptedBatch
        buffer.removeFirst(count)

        guard let events = try? batch.map({ try JSONSerialization.jsonObject(with: $0.data) }),
              let body = try? JSONSerialization.data(withJSONObject: ["batch": events]) else {
            completion?()
            return
        }
        var request = URLRequest(url: config.endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.setValue("Bearer \(config.projectKey)", forHTTPHeaderField: "authorization")
        request.httpBody = body

        let task = session.dataTask(with: request) { [weak self] data, response, error in
            guard let self else {
                completion?()
                return
            }
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            let retryBatch: [QueuedEvent]
            if error != nil || status == 408 || status == 429 || status >= 500 {
                retryBatch = batch
            } else if status >= 400 {
                retryBatch = []
            } else {
                let eventIDs = batch.compactMap { queuedEvent -> String? in
                    guard let event = try? JSONSerialization.jsonObject(with: queuedEvent.data) as? [String: Any] else {
                        return nil
                    }
                    return event["event_id"] as? String
                }
                if let data,
                   let retryIDs = abtoRetryEventIDs(responseData: data, eventIDs: eventIDs) {
                    retryBatch = batch.filter { queuedEvent in
                        guard let event = try? JSONSerialization.jsonObject(with: queuedEvent.data) as? [String: Any],
                              let eventID = event["event_id"] as? String else {
                            return true
                        }
                        return retryIDs.contains(eventID)
                    }
                } else {
                    retryBatch = batch
                }
            }
            self.finish(retryBatch: retryBatch, completion: completion)
        }
        task.resume()
    }

    private func finish(retryBatch: [QueuedEvent], completion: (@Sendable () -> Void)?) {
        queue.async {
            let now = Date()
            let eligible = retryBatch.filter {
                abtoRetryEligible(attempts: $0.attempts, firstQueuedAt: $0.firstQueuedAt, now: now)
            }
            if !eligible.isEmpty {
                // Cap the buffer so a dead endpoint cannot grow memory without bound.
                self.buffer = Array((eligible + self.buffer).prefix(abtoMaxBufferedEvents))
                let attempt = eligible.map(\.attempts).max() ?? 1
                let exponential = min(abtoMaxRetryDelay, self.config.flushInterval * pow(2.0, Double(attempt - 1)))
                let jitter = Double.random(in: 0..<max(0.001, exponential * abtoRetryJitterRatio))
                self.armTimerLocked(delay: exponential + jitter)
            } else if !self.buffer.isEmpty {
                self.flushLocked()
            }
            completion?()
        }
    }

}
